### PR TITLE
Relationships networking

### PR DESCRIPTION
**Pull request by Shatur**
Implements #340. I decided to name the method `replicate_together` instead of `replicate_related` to avoid confusion with the existing `replicate` method. The implementation properly handles all potential relationship edge cases including intersections, overlaps, splitting, and joining.
I also added unit tests to properly check message packing. In fact, most of the added lines are tests :sweat_smile:

### CODE REVIEW

**Review by UkoeHB — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Added
 
 - Support for `no_std`.
+- Relationships networking. Use `ReplicateTogetherAppExt::replicate_together<C>` to ensure that entities related by `C` are replicated in sync.
```
[the remark this answers is not in the document]
Comment (Shatur) [reply CHANGELOG.md:13]: I think using `replicate_*` naming could be confusing... We have `replicate` and `replicate_with` that register components for replication, But `replicate_together` works differently - it doesn't actually enable replication for `C`, but rather groups entities if they have this relationship.

File: src/shared/replication/related_entities.rs [hunk 2]
```diff
@@ -0,0 +1,585 @@
...
+        self.graph.remove_node(node);
+        self.entity_to_node.remove(&entity);
+        self.node_to_entity.remove(&node);
+    }
+
+    /// Recalculates graphs from SCC if there were any changes.
+    ///
+    /// The recalculation is not incremental, so it isn't performed automatically
+    /// on every change. Instead, manually call this before replication begins.
+    pub(crate) fn rebuild_graphs(&mut self) {
```
[the remark this answers is not in the document]
Comment (Shatur) [reply src/shared/replication/related_entities.rs:169]: Makes total sense, working on the benchmark to measure the impact first.

File: src/shared/replication/related_entities.rs [hunk 3]
```diff
@@ -0,0 +1,585 @@
...
+        self.entity_graphs.get(&entity).copied()
+    }
+
+    pub(crate) fn graphs_count(&self) -> usize {
+        self.graphs_count
+    }
+}
+
+fn add_relation<C: Relationship>(
+    trigger: Trigger<OnInsert, C>,
```
[the remark this answers is not in the document]
Comment (Shatur) [reply src/shared/replication/related_entities.rs:205]: Good point! I wonder how Bevy keeps `Children` updated when `Parent` changes :thinking:

File: src/server/replication_messages/mutations.rs [hunk 4]
```diff
@@ -59,55 +52,84 @@ impl Mutations {
...
+                EntityLocation::Related { index } => self.related[index].last(),
+                EntityLocation::Standalone => self.standalone.last(),
+            })
+            .map(|mutations| &mutations.ranges)
     }
 
     /// Removes last added entity from [`Self::add_entity`] with associated components.
     ///
     /// keeps allocated memory for reuse.
     pub(super) fn pop(&mut self) {
```
[the remark this answers is not in the document]
Comment (Shatur) [reply src/server/replication_messages/mutations.rs:119]: Good catch, added `take()`.

File: src/shared/replication/related_entities.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (UkoeHB) [reply src/shared/replication/related_entities.rs:205]: `ChildOf` is an immutable component. They use OnInsert/OnReplace.

[the remark this answers is not in the document]
Comment (Shatur) [reply src/shared/replication/related_entities.rs:205]: Ah, right! Let's restrict to immutable components then.

File: CHANGELOG.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (UkoeHB) [reply CHANGELOG.md:13]: The english meaning of `replicate_together` implies you are giving the function 2 things that should replicate together. What about `sync_related_entities::<C>`?

[the remark this answers is not in the document]
Comment (Shatur) [reply CHANGELOG.md:13]: Ah, makes sense!
I agree on `sync_related_entities::<C>`. A few more options to consider:
- `replicate_in_group`
- `sync_related` (a bit shorter, I think entities are implied)
- `replicate_in_sync`

Which name would you use?

[the remark this answers is not in the document]
Comment (UkoeHB) [reply CHANGELOG.md:13]: I like `sync_related_entities`. `sync_related` is slightly confusing because you are giving it a component type. The reader shouldn't have to think hard to figure out its meaning. The `replicate_*` variants are no good because the function doesn't replicate `C`.

[the remark this answers is not in the document]
Comment (Shatur) [reply CHANGELOG.md]: Agree, done.

File: src/shared/replication/related_entities.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Shatur) [reply src/shared/replication/related_entities.rs:179]: Added the performance mention. I will push benchmarks later.
Maybe could be a follow-up to simplify the review.

**Review by UkoeHB — APPROVED** [review]
